### PR TITLE
Fix #7375: Lazy Load Tabs when opening batch URLs

### LINKED ISSUE #7375: Opening/Closing a lot of tabs causes UI stutter
### Description: 
- Opening or Closing a lot of tabs, causes the UI to freeze or stutter.

### Steps to Reproduce 
  1. Bookmark 10+ tabs
  2. Hold down the finger on the bookmarks folder and choose "Open All"
  3. All tabs open

### Actual result: 
- UI freeze

### Expected result:
- Should lazy load and open all tabs without issue

### Reproduces how often: [Easily reproduced, Intermittent Issue]
- Easily reproduced

### Brave Version: 
- 1.51 (actually not sure???)

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary of Changes
- Lazy Load Tabs when opening a lot of them (I chose 5 arbitrarily because my 2015 MBP simulator lags with more than 5)
- Optimize restoring tabs.
- Optimize recently closed tabs.

This pull request fixes #7375

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/Brave/Frontend/Browser/TabManager.swift [hunk 1]
```diff
@@ -416,12 +416,30 @@ class TabManager: NSObject {
...
-    for url in urls {
-      tab = self.addTab(URLRequest(url: url), flushToDisk: false, zombie: zombie, isPrivate: isPrivate)
+    let tabs = urls.map {
+      let tab = self.addTab(URLRequest(url: $0),
+                            flushToDisk: false,
+                            zombie: true,
+                            isPrivate: isPrivate)
+      tab.lastTitle = $0.absoluteDisplayString
+      tab.url = $0
+      tab.favicon = FaviconFetcher.getIconFromCache(for: $0) ?? Favicon.default
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/TabManager.swift]: Should we lazy-load favicons as well?

File: Sources/Brave/Frontend/Browser/TabManager.swift [hunk 2]
```diff
@@ -416,12 +416,30 @@ class TabManager: NSObject {
     }
     // When bulk adding tabs don't notify delegates until we are done
     self.isRestoring = true
-    var tab: Tab!
-    for url in urls {
-      tab = self.addTab(URLRequest(url: url), flushToDisk: false, zombie: zombie, isPrivate: isPrivate)
+    let tabs = urls.map {
+      let tab = self.addTab(URLRequest(url: $0),
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/TabManager.swift]: Its a bit weird to have side-effects (addTab which alters the state of the class) trigger inside of a `map` which is just supposed to transform data from 1 type to another

File: Sources/Brave/Frontend/Browser/TabManager.swift [hunk 3]
```diff
@@ -416,12 +416,30 @@ class TabManager: NSObject {
...
+      tab.favicon = FaviconFetcher.getIconFromCache(for: $0) ?? Favicon.default
+      return tab
+    }
+    
+    // Load at most X of the most recent tabs
+    // IE: Load the last X tabs, lazy load all the rest
+    let amountOfTabsToRestoreImmediately = 5
+    
+    let urls = Array(urls.suffix(amountOfTabsToRestoreImmediately))
+    tabs.suffix(amountOfTabsToRestoreImmediately).enumerated().forEach { index, tab in
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/TabManager.swift]: I assume we dont have to worry about ordering?

File: Sources/Brave/Frontend/Browser/TabManager.swift [hunk 4]
```diff
@@ -938,15 +954,45 @@ class TabManager: NSObject {
...
   func restoreTab(_ tab: Tab) {
-    // Tab was created with no active webview or session data. Restore tab data from CD and configure.
-    guard let sessionTab = SessionTab.from(tabId: tab.id) else { return }
-
-    if !sessionTab.interactionState.isEmpty {
-      if let webView = tab.webView {
-        tab.navigationDelegate = navDelegate
-        tab.restore(webView, restorationData: (sessionTab.title, sessionTab.interactionState))
+    guard let webView = tab.webView else { return }
+    guard let sessionTab = SessionTab.from(tabId: tab.id) else {
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/TabManager.swift:969]: These can be combined into 1 guard

File: Sources/Brave/Frontend/Browser/TabManager.swift [hunk 5]
```diff
@@ -1054,29 +1100,27 @@ class TabManager: NSObject {
...
     }
     
     return SavedRecentlyClosed(
       url: fetchedTabURL,
-      title: fetchedTab.title,
-      interactionState: fetchedTab.interactionState,
-      order: fetchedTab.index)
+      title: tab.displayTitle,
+      interactionState: interactionState,
+      order: -1)
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/TabManager.swift:1134]: -1?

File: Sources/Data/models/SessionTab.swift [hunk 6]
```diff
@@ -193,27 +203,29 @@ extension SessionTab {
...
-          try context.save()
-        } catch {
-          Logger.module.error("performTask save error: \(error.localizedDescription, privacy: .public)")
-        }
+    DataController.perform { context in
+      guard !SessionTab.exists(tabId: tabId, in: context) else {
+        return
+      }
+      
+      guard let window = SessionWindow.getActiveWindow(context: context) else { return }
```
Comment (kylehickinson) [inline Sources/Data/models/SessionTab.swift]: These guards can be merged

File: Sources/Brave/Frontend/Browser/TabManager.swift (hunk 1 above)
Comment (Brandon-T) [reply Sources/Brave/Frontend/Browser/TabManager.swift]: I found it weird to not show them. Reason is because then I see blank tabs with blank fav-icons (the globe/default icon)

File: Sources/Brave/Frontend/Browser/TabManager.swift (hunk 5 above)
Comment (Brandon-T) [reply Sources/Brave/Frontend/Browser/TabManager.swift:1134]: Yeah because the `RecentlyClosed.insertAll` does the re-ordering of them all.

File: Sources/Brave/Frontend/Browser/TabManager.swift [hunk 7]
```diff
@@ -416,10 +416,27 @@ class TabManager: NSObject {
...
+    for (index, url) in urls.enumerated() {
+      tab = self.addTab(URLRequest(url: url),
+                            flushToDisk: false,
+                            zombie: true,
+                            isPrivate: isPrivate)
+      tab.lastTitle = url.absoluteDisplayString
+      tab.url = url
+      tab.favicon = FaviconFetcher.getIconFromCache(for: url) ?? Favicon.default
+      
+      if index >= urls.count - 1 - amountOfTabsToRestoreImmediately {
```
Comment (kylehickinson) [inline Sources/Brave/Frontend/Browser/TabManager.swift]: This reads a bit strange, I assume this is loading the _last_ 5 items in the `urls` list? Are these already ordered correctly?

Comment (Brandon-T) [reply Sources/Brave/Frontend/Browser/TabManager.swift]: This loads the last 5 items yeah. When we pass in the `urls` they are expected to already be in the correct order. This function only gets called from `Bookmarks` when you open all bookmarks for a folder at once.
So if a folder contains 10 bookmarks, it will open them in that exact order when you long-press on the folder and hit `Open All (10)`.

It will Lazy Load the last 5, and immediately load the first 5.
So when the user opens the tab tray, they will see the 5 pages loaded. If they scroll up, the others will not be loaded yet until they tap on them. Tabs are ordered top down in tab-tray.

**TLDR**: It loads them in the reverse order of the top 5. I could technically add them all to an array, suffix the last 5, and load exact.

**EDIT**: Made the code much clearer. I just added them all to an array, reverse the order of the top 5 that are immediately loaded, so the first tab the user sees is loaded first.

**Review by kylehickinson — APPROVED** [review]

**Review by soner-yuksel — APPROVED** [review]
